tensorflow/compiler/mlir/hlo/lib/Transforms/copy_removal.cc

```diff
@@ -0,0 +1,186 @@
+    SmallPtrSet<Value, 16U> processedAliases;
+
+    // Iterate over the allocs.
+    for (const BufferPlacementAllocs::AllocEntry &entry : allocs) {
+      Value allocValue = std::get<0>(entry);
+
+      // Resolve the aliases of the current alloc and iterate over them.
+      const ValueSetT &aliasSet = aliases.resolve(allocValue);
+      for (Value alias : aliasSet) {
+        // If the alias is already processed, continue.
+        if (!processedAliases.insert(alias).second)
+          continue;
+
+        // If the value has no uses/ empty userange, continue.
+        auto userangeInterval = userange.getUserangeInterval(alias);
+        if (!userangeInterval)
+          continue;
+
+        // Iterate over the UseIntervals and check if the last Operation in the
+        // UseInterval implements a CopyOpInterface.
+        for (UseInterval interval : *userangeInterval.getValue()) {
+          Operation *currentLastUse = userange.getOperation(interval.end);
+          auto copyOpInterface = dyn_cast<CopyOpInterface>(currentLastUse);
+          if (!copyOpInterface)
+            continue;
+
+          // Check if the source is the alias.
+          if (copyOpInterface.getSource() != alias)
+            continue;
+
+          toProcess.insert(currentLastUse);
+        }
+      }
+    }
+  }
+
+  /// Check if all uses of the target Value are dominated by given Operation.
+  /// Note: The target has always at least one use which is the copy operation.
+  bool checkDominance(Operation *useOp, Value target) {
+    Block *useBlock = useOp->getBlock();
+    const UserangeAnalysis::UsePositionList &targetUsePosList =
+        *userange.getUserangePositions(target).getValue();
+    // Check if any use of the target is not dominated by the useOp. Erased
+    // operations are ignored as uses.
+    return !llvm::any_of(targetUsePosList,
+                         [=](const UserangeAnalysis::UsePosition targetUsePos) {
+                           Operation *targetUse = targetUsePos.second;
+                           return toErase.find(targetUse) == toErase.end() &&
+                                  !dominators.dominates(useBlock,
+                                                        targetUse->getBlock());
+                         });
+  }
+
+  /// Check if the given Operation is an ancestor of the operations inside the
+  /// UseInterval.
+  bool checkAncestor(Operation *op, UseInterval &interval) {
+    // Divide the start and end by two to remove read/write properties.
+    for (int id = interval.start / 2, e = interval.end / 2; id <= e; ++id) {
+      // Get the operation from the id. Multiply the id by 2, because the
+      // userange operates on doubled ids. Return false if the operation is not
+      // an ancestor.
+      if (!op->isAncestor(userange.getOperation(id * 2)))
+        return false;
+    }
+    return true;
+  }
+
+  /// A set containing copy operations that can be erased.
+  SmallPtrSet<Operation *, 16> toErase;
```

Comment:
This should not be a field. Instead make it a local and pass it explicitly into `checkDominance` maybe as an `ignore_set` parameter.